eventsourcing/src/test/java/org/axonframework/eventsourcing/AnnotationBasedEventStateApplierTest.java

```diff
@@ -0,0 +1,343 @@
+
+            // then
+            assertEquals("null-0-1-2", state.handledPayloads);
+            assertEquals(3, state.handledCount);
+        }
+    }
+
+    @Nested
+    class ParameterResolution {
+
+        @Test
+        void resolvesMetadata() {
+            // given
+            var state = new TestState();
+            var event = domainEvent(0, "sampleValue");
+
+            // when
+            state = eventStateApplier.apply(state, event, processingContext);
+
+            // then
+            assertEquals("null-sampleValue", state.handledMetadata);
+        }
+
+        @Test
+        void resolvesSequenceNumber() {
+            // given
+            var state = new TestState();
+            var event = domainEvent(0);
+
+            // when
+            state = eventStateApplier.apply(state, event, processingContext);
+
+            // then
+            assertEquals("null-0", state.handledSequences);
+        }
+
+        @Test
+        void resolvesSources() {
+            // given
+            var state = new TestState();
+            var event = domainEvent(0);
+
+            // when
+            state = eventStateApplier.apply(state, event, processingContext);
+
+            // then
+            assertEquals("null-id", state.handledSources);
+        }
+
+        @Test
+        void resolvesTimestamps() {
+            var timestamp = Instant.now();
+            GenericEventMessage.clock = Clock.fixed(timestamp, ZoneId.systemDefault());
+
+            // given
+            var state = new TestState();
+            var event = domainEvent(0);
+
+            // when
+            state = eventStateApplier.apply(state, event, processingContext);
+
+            // then
+            assertEquals("null-" + timestamp, state.handledTimestamps);
+        }
+    }
+
+    @Nested
+    class HandlerInvocationRules {
+
+        @Test
+        void doNotHandleNotDeclaredEventType() {
+            // given
+            var eventStateApplier = new AnnotationBasedEventStateApplier<>(HandlingJustStringState.class);
+            var state = new HandlingJustStringState();
+            var event = domainEvent(0);
+
+            // when
+            state = eventStateApplier.apply(state, event, processingContext);
+
+            // then
+            assertEquals(0, state.handledCount);
+        }
+
+        @Test
+        void invokesOnlyMostSpecificHandler() {
+            // given
+            var state = new TestState();
+            var event = domainEvent(0);
+
+            // when
+            state = eventStateApplier.apply(state, event, processingContext);
+
+            // then
+            assertEquals("null-0", state.handledPayloads);
+            assertFalse(state.objectHandlerInvoked);
+            assertEquals(1, state.handledCount);
+        }
+    }
+
+    // feel free to drop the support in case of some changed in EventHandling logic (like Empty stream always returned from event handling component)
+    @Nested
```

Comment:
Not something to adjust for this PR, but a "brainfart" to support this.
I think we have roughly two options to support the state object from an event sourcing handler, being:

1. Make an `EventSourcingHandler` interface that supports `MessageStream.Single` of a state-holding `Message` implementation.
2. Add the state in the `Context` of the `MessageStream.Entry` instead.

So, some food for thought per chance! I would like this to work eventually too actually. :-)

Reply:
I started a new discussing around that here:
https://github.com/AxonFramework/AxonFramework/pull/3308#discussion_r1967662537

Reply:
Awesome, thanks for that!